C#9 pattern enhancements

gerhardol opened the pull request:
Fixes #8728
Separated from #8734 (based on this)

## Proposed changes

Enable C#9 'syntactic sugar' for pattern matching.
Mostly changed with intellicode suggestions and a few manual changes.

Separated from #8734 to merge that quickly and discuss the changes here.
I would like to apply the changes in FileViewer.cs at least.
Maybe StyleCop changes are required toadd parenthesis to improve readability to merge this.

## Test methodology 

manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie reviewed (commented):
I'm not going to block this work, but so far this syntax makes me think more. Perhaps with time it'll get easier.

pmiossec commented:
No problem for me to read it. I like it.
But I think it will take a lot of time to think about using it...

On Tue, Jan 12, 2021, 01:03 Igor Velikorossov <notifications@github.com>
wrote:

> @RussKie <https://github.com/RussKie> requested review from
> @gitextensions/git-extensions-team on: #8737
> <https://github.com/gitextensions/gitextensions/pull/8737> C#9 pattern
> enhancements.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8737#event-4193340165>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQLJDGZXRFB2L7VUS7JXTSZOGW3ANCNFSM4V4KI3AA>
> .
>

gerhardol commented:
Rebased to resolve conflicts
An alternative supported by StyleCop is the following. 

Before (slightly tweaked)
```C#
        public static bool IsValidChar(char c)
        {
            return c > ' ' && c < '~' && c != '^' && c != ':' &&
                    Array.IndexOf(Path.GetInvalidPathChars(), c) < 0;
        }

```

Currently in the PR
```C#
        public static bool IsValidChar(char c)
        {
            return (c is > ' ' and < '~' and not '^' and not ':') &&
                    Array.IndexOf(Path.GetInvalidPathChars(), c) < 0;
        }
```

Alternative (must be is())
```C#
        public static bool IsValidChar(char c)
        {
            return c is(> ' ' and < '~' and not '^' and not ':') &&
                    Array.IndexOf(Path.GetInvalidPathChars(), c) < 0;
        }
```

RussKie commented on ResourceManager/GitExtensionsControl.cs:
```diff
@@ -123,7 +123,7 @@ public static bool IsTextEditKey(Keys keys, bool multiLine = false)
             switch (keys)
             {
-                case Keys key when (key >= Keys.A && key <= Keys.Z) || (key >= Keys.D0 && key <= Keys.D9) || (key >= Keys.Oem1 && key <= Keys.Oem102):
+                case Keys key when key is >= Keys.A and <= Keys.Z or >= Keys.D0 and <= Keys.D9 or >= Keys.Oem1 and <= Keys.Oem102:
```
How about this?
```diff
diff --git a/ResourceManager/GitExtensionsControl.cs b/ResourceManager/GitExtensionsControl.cs
index 2c23fc061..f2301ec8c 100644
--- a/ResourceManager/GitExtensionsControl.cs
+++ b/ResourceManager/GitExtensionsControl.cs
@@ -123,7 +123,10 @@ public static bool IsTextEditKey(Keys keys, bool multiLine = false)
             keys &= ~Keys.Shift; // ignore the SHIFT key as modifier
             switch (keys)
             {
-                case Keys key when key is >= Keys.A and <= Keys.Z or >= Keys.D0 and <= Keys.D9 or >= Keys.Oem1 and <= Keys.Oem102:
+                case Keys key when key is
+                    (>= Keys.A and <= Keys.Z) or
+                    (>= Keys.D0 and <= Keys.D9) or
+                    (>= Keys.Oem1 and <= Keys.Oem102):
                 case Keys.Space:
                 case Keys.Back:
                 case Keys.Delete:
```

sharwell replied:
Considering what the code on the left is, I'm fine with keeping one line as long as the parentheses are returned. If wrapping is applied, I would suggest placing the `or` operators at the beginning of lines instead of the end.

RussKie replied:
> suggest placing the `or` operators at the beginning of lines instead of the end.

Works for me

RussKie commented on GitUI/Script/SplitButton.cs:
```diff
@@ -321,7 +321,7 @@ protected override void OnPaint(PaintEventArgs pevent)
                               bounds.Height - (internalBorder * 2) + 2);
 
-            bool drawSplitLine = State == PushButtonState.Hot || State == PushButtonState.Pressed || !Application.RenderWithVisualStyles;
+            bool drawSplitLine = State is PushButtonState.Hot or PushButtonState.Pressed || !Application.RenderWithVisualStyles;
```
```suggestion
            bool drawSplitLine = !Application.RenderWithVisualStyles || (State is PushButtonState.Hot or PushButtonState.Pressed);
```

sharwell replied:
Intensely agree with the suggestion here. Parentheses are essential.

gerhardol replied:
Added parenthesis in general. Did not change the order here (kept as below)

```C#
bool drawSplitLine = State is(PushButtonState.Hot or PushButtonState.Pressed) || !Application.RenderWithVisualStyles;
```

RussKie commented on GitUI/GitUICommands.cs:
```diff
@@ -1431,7 +1431,7 @@ public bool RunCommand(IReadOnlyList<string> args)
             }
 
-            if ((command == BlameHistoryCommand || command == FileHistoryCommand) && args.Count <= 2)
+            if ((command is BlameHistoryCommand or FileHistoryCommand) && args.Count <= 2)
```
```suggestion
            if (args.Count <= 2 && (command is BlameHistoryCommand or FileHistoryCommand))
```

sharwell replied:
I prefer the form already in code. The condition which restricts the argument count is the command, so it should be checked first unless fast-path performance dictates otherwise (which is not a problem in this case).

RussKie reviewed (commented):
I'm not a big fan, but I can live with this

sharwell commented on GitCommands/Git/GitBranchNameNormaliser.cs:
```diff
@@ -107,8 +107,7 @@ public string Normalise(string branchName, GitBranchNameOptions options)
         {
-            return (c >= 32 && c < 127) &&
-                    c != ' ' && c != '~' && c != '^' && c != ':' &&
+            return (c is > ' ' and < '~' and not '^' and not ':') &&
```
💭 No way I would have known that `'~'` is 0x7E. We should document why this range has been chosen.

gerhardol replied:
This is documented in the top of the file. I updated the link and compared to current (adding one rule).

sharwell commented on Plugins/Statistics/GitStatistics/PieChart/EdgeColor.cs:
```diff
@@ -51,7 +51,7 @@ public static Color GetRenderingColor(EdgeColorType edgeColorType, Color color)
         private static EdgeColorType GetContrastColorType(Color color, EdgeColorType colorType)
         {
-            Debug.Assert(colorType == EdgeColorType.Contrast || colorType == EdgeColorType.EnhancedContrast, "colorType == EdgeColorType.Contrast || colorType == EdgeColorType.EnhancedContrast");
+            Debug.Assert(colorType is EdgeColorType.Contrast or EdgeColorType.EnhancedContrast, "colorType == EdgeColorType.Contrast || colorType == EdgeColorType.EnhancedContrast");
```
```suggestion
            Debug.Assert(colorType is EdgeColorType.Contrast or EdgeColorType.EnhancedContrast, "colorType is EdgeColorType.Contrast or EdgeColorType.EnhancedContrast");
```

sharwell commented on Plugins/Statistics/GitStatistics/PieChart/PieChart3D.cs:
```diff
@@ -851,7 +851,7 @@ protected virtual void InitializePieSlices()
         protected SizeF GetSliceDisplacement(float angle, float displacementFactor)
         {
-            Debug.Assert(displacementFactor > 0F && displacementFactor <= 1F, "displacementFactor > 0F && displacementFactor <= 1F");
+            Debug.Assert(displacementFactor is > 0F and <= 1F, "displacementFactor > 0F && displacementFactor <= 1F");
```
```suggestion
            Debug.Assert(displacementFactor is > 0F and <= 1F, "displacementFactor is > 0F and <= 1F");
```

gerhardol replied:
ok

sharwell commented on Plugins/Statistics/GitStatistics/PieChart/PieChart3D.cs:
```diff
@@ -851,7 +851,7 @@ protected virtual void InitializePieSlices()
         protected SizeF GetSliceDisplacement(float angle, float displacementFactor)
         {
-            Debug.Assert(displacementFactor > 0F && displacementFactor <= 1F, "displacementFactor > 0F && displacementFactor <= 1F");
+            Debug.Assert(displacementFactor is > 0F and <= 1F, "displacementFactor > 0F && displacementFactor <= 1F");
```
💡 Consider using `IsDisplacementValid` instead of duplicating the check.

```suggestion
            Debug.Assert(IsDisplacementValid(displacementFactor), "IsDisplacementValid(displacementFactor)");
```

gerhardol replied:
IsDisplacementValid() is not the same but ==0F is handled anyway so I will change

sharwell reviewed (commented):
Need documentation for one of the changes, and parentheses added for cases where `and` and `or` are both used, and cases where `and`/`or` are used with `&&`/`||`.

gerhardol commented:
Added parenthesis to all changes (as it was only statements with && or || that was changed).
I was not completely sure what paren style that was suggested, I added my last suggestion, also suggested here:
https://github.com/gitextensions/gitextensions/pull/8748#discussion_r556874914
That can be changed in this PR (and StyleCop should allow a space after is before paren).

RussKie commented:
@sharwell do you think it is feasible to enhance [SA1008](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1008.md) to play nicely with the new syntax?
Do you know if this issue is tracked anywhere, or do I need to raise a new issue?

sharwell commented:
Can you file a new issue for it?

RussKie commented:
@gerhardol can you try updating the package to the latest beta https://www.nuget.org/packages/StyleCop.Analyzers?

gerhardol commented:
Updated to latest beta package of StyleCop. We were already using a beta version, so not a big difference
Also rewrote a few statements
One new check had to be addressed too

Separate commits, for reviews

RussKie commented:
```
1) Error : GitUITests.UserControls.RevisionGrid.RevisionGraphMultiThreadingTests.ShouldReorderInTopoOrder
  System.AggregateException : One or more errors occurred.
    ----> System.ArgumentException : Unable to sort because the IComparer.Compare() method returns inconsistent results. Either a value does not compare equal to itself, or one value repeatedly compared to another value yields different results. IComparer: 'System.Array+FunctorComparer`1[GitUI.UserControls.RevisionGrid.Graph.RevisionGraphRevision]'.
     at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
     at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout)
     at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
     at GitUITests.UserControls.RevisionGrid.RevisionGraphMultiThreadingTests.ShouldReorderInTopoOrder() in C:\projects\gitextensions\UnitTests\GitUI.Tests\UserControls\RevisionGrid\Graph\RevisionGraphMultiThreadingTests.cs:line 62
  --ArgumentException
     at System.Collections.Generic.IntrospectiveSortUtilities.ThrowOrIgnoreBadComparer(Object comparer)
     at System.Collections.Generic.ArraySortHelper`1.Sort(T[] keys, Int32 index, Int32 length, IComparer`1 comparer)
     at System.Array.Sort[T](T[] array, Int32 index, Int32 length, IComparer`1 comparer)
     at System.Collections.Generic.List`1.Sort(Comparison`1 comparison)
     at GitUI.UserControls.RevisionGrid.Graph.RevisionGraph.BuildOrderedNodesCache(Int32 currentRowIndex) in C:\projects\gitextensions\GitUI\UserControls\RevisionGrid\Graph\RevisionGraph.cs:line 336
     at GitUI.UserControls.RevisionGrid.Graph.RevisionGraph.GetNodeForRow(Int32 row) in C:\projects\gitextensions\GitUI\UserControls\RevisionGrid\Graph\RevisionGraph.cs:line 130
     at GitUITests.UserControls.RevisionGrid.RevisionGraphMultiThreadingTests.Render() in C:\projects\gitextensions\UnitTests\GitUI.Tests\UserControls\RevisionGrid\Graph\RevisionGraphMultiThreadingTests.cs:line 113
     at GitUITests.UserControls.RevisionGrid.RevisionGraphMultiThreadingTests.<ShouldReorderInTopoOrder>b__5_2() in C:\projects\gitextensions\UnitTests\GitUI.Tests\UserControls\RevisionGrid\Graph\RevisionGraphMultiThreadingTests.cs:line 56
     at System.Threading.Tasks.Task.InnerInvoke()
     at System.Threading.Tasks.Task.Execute()
```

gerhardol commented:
> 1) Error : GitUITests.UserControls.RevisionGrid.RevisionGraphMultiThreadingTests.ShouldReorderInTopoOrder

Seem to be a flaky test. I do not see any of my changes related to this and it runs OK locally.

gerhardol commented:
@msftbot merge in 24 hours
